### ISSUE TITLE
add: ~2.5X slower than 1.11.15 for a directory

### ISSUE THREAD

**Issue by skshetry** [iterative/dvc#5746]
### How to reproduce
```bash
# initialize
cd "$(mktemp -d)"
git init && dvc init

# prepare data
mkdir data
for i in $(seq 100000); do
  echo foo-$i > data/$i.txt
done

pip install dvc==1.11.15
dvc --version
dvc add data --cprofile --cprofile-dump old-dump.prof > /dev/null

# update to the latest version
pip install -U dvc
dvc --version
dvc add data --cprofile-dump new-dump.prof > /dev/null
```

There are a few problems here:
1. The progress bar flickers at the end. We are creating a progress bar for moving a .tmp file, which makes progress bar flicker. But this also means that we are spending too much time in creating progress bar, which is almost ~10% of runtime.
 In 1.11.X, I don't notice that kind of progress bars, but needs a closer look.
2. `md5` computation is almost 2.5X slow than 1.11.X. It never even hits close to the average in 1.11 version.
3. It seems that it's also trying to checkout files back to the workspace, but I am not quite familiar with `dvc.objects`, so I could be wrong here. 

https://github.com/iterative/dvc/blob/4d4f9427d1e57fd7c3db68965e41d95d2b197339/dvc/output/base.py#L318-L338

I'll try to gather more information later. Just writing it out what I saw right now.

1.11.16 yappi log: 
[old.log](https://github.com/iterative/dvc/files/6262021/old.log)

`master` yappi log: 
[new.log](https://github.com/iterative/dvc/files/6262023/new.log)

**Comment by skshetry** [comment]
Looks like this was correctly reported by `dvc-bench` on https://github.com/iterative/dvc-bench/issues/239 (and, thanks to bisecting from @pared, points to #5505 as the actual cause).

![Screenshot from 2021-04-05 17-18-41](https://user-images.githubusercontent.com/18718008/113569924-0a957e80-9633-11eb-8ddf-dc764f9e9333.png)

**Comment by shcheklein** [comment]
It's interesting to see also who it had been degrading from 30.0 to 45.0 before that abrupt regression. I wonder if we had any notification about that and what was the cause (can be just time it takes to run the command)

**Comment by skshetry** [comment]
The long small bumps from 0.90.0 to 1.11.15, with ~31s -> ~36s, I can only speculate it to:
1. respecting .dvcignore in more places
2. trying to avoid dvc subrepos
3. recollecting stages multiple times to make the graph consistent

Maybe few other things that were added.

The other bumps between 1.11.15 and 2.0.0a0 seem to be because of `dvc.objects` refactoring (went from ~36s to ~44s).
The major regression seems to be because of the `dvc.state` changes on #5505 (from ~44s to ~86s).

I did compare benchmark results between current `object.stage()` and `get_dir_hash` (old method) and it points to three things being the culprit:

1. Saving md5s with the new state is a lot slower.
2. We started using LocalFileSystem when collecting hashes, which respects dvcignore. Before, we used to directly call `os.path`.
3. Lots of file system calls than before.

(PTAL at the profiling log attached above for more)
